Add generic join unit tests

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61786 Add generic join unit tests**
* #61785 Minor documentation fixes
* #61784 Fix indent

This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

Commit:
Add generic join unit tests

[ghstack-poisoned]

Commit:
Update on "Add generic join unit tests"




[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29746646).

mrshenli reviewed (commented):
Do you need to add this new test file to https://github.com/pytorch/pytorch/blob/master/test/run_test.py to enable it?

Commit:
Update on "Add generic join unit tests"


This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29746646).

mrshenli commented on test/distributed/algorithms/test_join.py:
```diff
@@ -0,0 +1,420 @@
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        with _Join([allreducer], run_post_hooks=True):
+            for _ in inputs:
+                allreducer()
```
Would I be correct if I assume main hook would still run in this case? And compared to `test_single_joinable`, the only difference here is that `test_single_joinable_post_hooks` does not check main hook results? If so, does it make sense if we remove this test, as it seems `test_single_joinable` behaves exactly the same way and only differs in asserts?

andwgu replied:
I have now made `test_single_joinable_post_hooks()` not run the main hooks.

mrshenli commented on test/distributed/algorithms/test_join.py:
```diff
@@ -0,0 +1,420 @@
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable(self):
```
does it make sense to dedup test code with `test_single_joinable_main_hooks`?

andwgu replied:
Yup, I factored the core logic out into a single `_test_join_base()` and now use that as the base for all tests.

mrshenli commented on test/distributed/algorithms/test_join.py:
```diff
@@ -0,0 +1,420 @@
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_multiple_joinables(self):
```
looks like this can also dedup with `test_single_joinable_main_hooks` by providing a list of hooks and expected results to a common function?

mrshenli commented on test/distributed/algorithms/test_join.py:
```diff
@@ -0,0 +1,420 @@
+        allreducer = AllReducer(self.device, self.process_group)
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with self.assertRaises(RuntimeError):
```
Shall we check the error message is expected?

andwgu replied:
Done.

mrshenli commented on test/distributed/algorithms/test_join.py:
```diff
@@ -0,0 +1,420 @@
+        ]
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with self.assertRaises(RuntimeError):
```
ditto

andwgu replied:
Done.

mrshenli reviewed (approved):
LGTM! I left some minor comments. Please make sure tests are launched in CI and tests pass before landing. 

When imported, please also add this to internal tests.

Commit:
Update on "Add generic join unit tests"


This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29746646).

Commit:
Update on "Add generic join unit tests"


This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29746646).

Commit:
Update on "Add generic join unit tests"


This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

[ghstack-poisoned]

Commit:
Update on "Add generic join unit tests"


This adds unit tests for the generic join context manager.

```
gpurun python test/distributed/algorithms/test_join.py
```

Differential Revision: [D29746646](https://our.internmc.facebook.com/intern/diff/D29746646)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29746646).
